{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook we take the orthologous sequences and perform MSA for each gene (using clustal omega). Then, we \"translate\" the alignment from amino-acids to nucleotides (using pal2nal). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import pickle\n",
    "import concurrent.futures\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import numpy as np\n",
    "import gzip\n",
    "import shutil\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_ortholog_sequences_to_fasta_single_gene(gene:str) -> None:\n",
    "    '''\n",
    "    This function takes the orthologs' data that is stored in dictionaries downloaded\n",
    "    from Ensembl using the REST API, and stores the aa and cdna sequences in fasta files\n",
    "    (We need the input of clustal omega as fasta files). \n",
    "    '''\n",
    "    try:\n",
    "        '''Needed paths'''\n",
    "        cdna_dict_path = f\"../co_trans_data/orthologs/cdna_dict/cdna_{gene}.pickle\" #a dict with the cdna sequences of the orthologs\n",
    "        aa_dict_path = f\"../co_trans_data/orthologs/aa_dict/aa_{gene}.pickle\" #a dict with the aa sequences of the orthologs\n",
    "        cdna_fasta_path = f\"../co_trans_data/orthologs/nt_before_msa/{gene}.fasta\" #where to write the fasta with the cdna sequences\n",
    "        aa_fasta_path = f\"../co_trans_data/orthologs/aa_before_msa/{gene}.fasta\" #where to write the fasta with the aa sequences\n",
    "\n",
    "        '''load the ortholog data dictionary'''\n",
    "        cdna_dict = pickle.load(open(cdna_dict_path, \"rb\"))\n",
    "        aa_dict = pickle.load(open(aa_dict_path, \"rb\"))\n",
    "\n",
    "        '''write fastas'''\n",
    "        #initilize files\n",
    "        nt_file = open(cdna_fasta_path, \"w\")\n",
    "        aa_file = open(aa_fasta_path, \"w\")\n",
    "        #add the human sequence\n",
    "        human_cdna_sequence = cdna_dict['data'][0]['homologies'][0]['source']['seq']\n",
    "        human_aa_sequence = aa_dict['data'][0]['homologies'][0]['source']['seq']\n",
    "        nt_file.write(f\">0_homo_sapiens\\n{human_cdna_sequence}\\n\")\n",
    "        aa_file.write(f\">0_homo_sapiens\\n{human_aa_sequence}\\n\")\n",
    "        #add the orthologous sequences\n",
    "        num_orthologs = len(cdna_dict['data'][0]['homologies'])\n",
    "        wanted_orthologs_counter = 1\n",
    "        for cur_ortholog in range(num_orthologs):\n",
    "            cdna_data = cdna_dict['data'][0]['homologies'][cur_ortholog]\n",
    "            aa_data = aa_dict['data'][0]['homologies'][cur_ortholog]\n",
    "            species_cdna = cdna_data ['target']['species']\n",
    "            species_aa = aa_data ['target']['species']\n",
    "            assert(species_cdna == species_aa)\n",
    "            cdna_seq = cdna_data['target']['seq']\n",
    "            aa_seq = aa_data['target']['seq']\n",
    "            perc_id = cdna_dict['data'][0]['homologies'][cur_ortholog]['target']['perc_id'] #percent of the orthologous sequence that matches the human sequence\n",
    "            if (len(cdna_seq) % 3 == 0) & (perc_id >= 50): #make sure that the sequence is a valid cds and that the identity is larger than 50\n",
    "                nt_file.write(f\">{wanted_orthologs_counter}_{species_cdna}\\n{cdna_seq}\\n\")\n",
    "                aa_file.write(f\">{wanted_orthologs_counter}_{species_aa}\\n{aa_seq}\\n\")\n",
    "                wanted_orthologs_counter += 1\n",
    "    except Exception as e:\n",
    "        print(f\"Could not perfrom this for gene {gene}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Get the list of needed genes'''\n",
    "genes_list = pd.read_pickle('../co_trans_data/unique_genes.pickle') #all protein-coding human genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Create the fasta files using parallel computing'''\n",
    "\n",
    "with concurrent.futures.ProcessPoolExecutor(max_workers=60) as executor:\n",
    "    futures = []\n",
    "    for gene in genes_list:\n",
    "        futures.append(executor.submit(write_ortholog_sequences_to_fasta_single_gene, gene=gene)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preform MSA using Clustal Omega"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_list = pd.read_pickle('../co_trans_data/genes_for_msa.pickle') #only genes with orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''use clustal omega [www.clustal.org] to perform the MSA'''\n",
    "\n",
    "def perform_msa(gene:str) -> None:\n",
    "    '''\n",
    "    This function calls Clustal Omega [www.clustal.org] from the CLI,\n",
    "    to perform the MSA on the amino acid orthologous sequences of a gene\n",
    "    '''\n",
    "    input_path = f'../co_trans_data/orthologs/aa_before_msa/{gene}.fasta'\n",
    "    output_path = f'../co_trans_data/orthologs/aa_after_msa/{gene}.fasta'\n",
    "    !clustalo -i $input_path -o $output_path #call clustal omega from the cli "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Create the MSAs using parallel computing'''\n",
    "\n",
    "with concurrent.futures.ProcessPoolExecutor(max_workers=50) as executor:\n",
    "    futures = []\n",
    "    for gene in genes_list:\n",
    "        futures.append(executor.submit(perform_msa, gene=gene)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change MSA from amino-acids to nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_list = pd.read_pickle('../co_trans_data/genes_for_msa.pickle') #only genes with orthologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Create batches of genes for parallalization'''\n",
    "num_wanted_cpus = 50\n",
    "num_genes = len(genes_list)\n",
    "num_genes_per_batch = int(np.round(num_genes /num_wanted_cpus))\n",
    "batches_of_genes = [genes_list[x:x+num_genes_per_batch] for x in range(0, num_genes, num_genes_per_batch)]\n",
    "#change the gene names in each batch to strings to fit the bash code\n",
    "batches_of_genes_string = []\n",
    "for batch in batches_of_genes:\n",
    "    batch = \":\".join(batch)\n",
    "    batches_of_genes_string.append(batch)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: pal2nal prints many warnings while running and there is no option to control the verbosity. \n",
    "This makes the notebook crash. So, instead of calling the cpus in parellel directly from the notebook as we did with clustalo, we will create a bash script that calls pal2nal from the cli. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting run_pal2nal.sh\n"
     ]
    }
   ],
   "source": [
    "%%writefile run_pal2nal.sh\n",
    "#!/bin/bash\n",
    "\n",
    "cd ./co_trans_code/\n",
    "\n",
    "eval \"$(conda shell.bash hook)\"\n",
    "conda activate co-trans\n",
    "\n",
    "cur_batch=$(echo $single_batch_of_genes|sed 's/:/ /g')\n",
    "\n",
    "for gene in $cur_batch;\n",
    "do\n",
    "aa_path=$aa_after_msa_path$gene\n",
    "nt_before_path=$nt_before_msa_path$gene\n",
    "nt_after_path=$output_path$gene\n",
    "\n",
    "pal2nal.pl \"${aa_path}.fasta\" \"${nt_before_path}.fasta\" -output fasta > \"${nt_after_path}.fasta\"\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many orthologs per gene?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = \"../co_trans_data/orthologs/aa_before_msa/\"\n",
    "\n",
    "num_orthologs = []\n",
    "for filename in os.listdir(directory):\n",
    "    \n",
    "    #unzip\n",
    "    with gzip.open(f\"{directory}{filename}\", 'rb') as f_in:\n",
    "        not_zipped_filename = \".\".join(filename.split(\".\")[:-1])\n",
    "        with open(f\"./{not_zipped_filename}_aa_before\", 'wb') as f_out:\n",
    "            shutil.copyfileobj(f_in, f_out)\n",
    "            \n",
    "    #count orthologs per gene\n",
    "    num = len([1 for line in open(f\"./{not_zipped_filename}_aa_before\") if line.startswith(\">\")])\n",
    "    num_orthologs.append(num)\n",
    "    os.remove(f\"./{not_zipped_filename}_aa_before\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Number of orthologs per gene')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEICAYAAABWJCMKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAY1klEQVR4nO3dfZRVV33G8e8jEIIhGDATxBnMkIovkGpeKGJjrUpqiDGS1dW046pmokS6Ir5Vqwvqu5Uaq8uadBmURgV8QxpNYSWNSlHUKA1OTCIhBBmFwAhhxqRUElsi+OsfZ485DpeZO8zLnTv7+ax11z33d84+Z+97mfvM2efORRGBmZnl50m17oCZmdWGA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOABswSaskfbhGx5akz0v6b0lbh+gYV0u64yTbfkDSFwe7T2aDwQEwCknaI+mgpNNKtWskba5ht4bKi4A/A5oiYu5AdyapWVJIGjvwrpmNbA6A0Wss8NZad6K/JI3pZ5OzgT0R8dggHNtv+iV+PkY/B8Do9THg7ySd0XNFpd9yJW2WdE1avlrSDyT9s6RDkn4u6Y9TfZ+kTkmtPXZ7pqSNkg5L+q6ks0v7fk5a94iknZL+srRulaQVkv5D0mPASyv09+mSNqT27ZLekOqLgJuAF0p6VNIHK7R9kqT3SHow9XuNpKf0eB4WSdoLfBv4Xmp6KO3zhaV9fTxNNe2WdGlf/atE0qskbU/P62ZJzy2tu0DS3ek5/DdJX+2eWpN0pqRbU7tHJH1fUsWf3zSmt6TX7ZeSPlbeVtLrJe1IY/lmj9cqJC2RtAvYdYL9X5Wez4clvTedcV5cer6XSvpZWr9O0pQez3erpL2pb+/u8VpVbGtDJCJ8G2U3YA9wMfB14MOpdg2wOS03AwGMLbXZDFyTlq8GjgKvA8YAHwb2Ap8CxgMvBw4DE9P2q9LjF6f11wN3pHWnAfvSvsYCFwC/BGaX2v4PcBHFLySnVhjPd4EbgVOB84AuYH6pr3f08ly8HmgHzgEmpufkCz2ehzWpnxNO8NxcDfwGeEN6Pq4F9gOqon8fAL6Ylp8FPEYxZTUOeFfq2ynp9iDFWds44M+Bx0uv30eAT6d144A/6T5+hTEH8B1gCvAM4Kel1/aKdMznptfjPcAPe7TdmNpOqLDvWcCjFFNvpwAfT8/NxWn924D/AprSv4XPAF/p8Xz/a3qunw8cAZ7bV1vfhui9otYd8G0IXtQnAuBcijfXBvofALtK6/4wbT+1VHsYOC8trwLWltZNBI4B04G/Ar7fo3+fAd5faruml7FMT/s6vVT7CLCq1NfeAmAT8MbS42enN6yxpefhnNL6Ss/N1UB76fGT0zZPq6J/H+CJAHgvsK603ZOAXwAvoQjPX1B6Uwfu4IkA+BCwHnhmFa9/AAtKj98IbErLtwOLevTh18DZpbYv62Xf7yu/Kafn4nGeCIAdpPBLj6dVeL6bSuu3Ai19ta31z9RovXkKaBSLiPuAW4GlJ9H8YGn5f9P+etYmlh7vKx33UeAR4OkUc/QvSFMXhyQdAv6a4s3zuLYVPB14JCIOl2oPAo1VjuPpafty27HA1CqP3+2h7oWI+HVanNjP/v1eXyLit+nYjWndLyK981Xo18cofnP/Vpra6es1Lbd9MO0fitfj+tJr8QigHv3t6/Uov9a/pvhloNvZwC2l/e+gCMjy8/1QafnXPPHvqJq2NogcAKPf+ymmLso/4N0XTJ9cqpXfkE/G9O4FSRMpphD2U7xZfDcizijdJkbEtaW2vX0l7X5giqTTS7VnUPy2XI39FG8s5bZH+f2AixMsV7v/avv3e32RJIrn7RfAAaAx1br97jmNiMMR8Y6IOAe4HHi7pPm99Gt6afkZ6dhQvB5/0+P1mBARPyxt39tzcIBiiqZ7DBOAp5bW7wMu7bH/UyOimtdrIG3tJDgARrmIaAe+CrylVOuieNN5jaQxkl4P/MEAD/UKSS+SdArwD8CdEbGP4gzkWZJeK2lcuv1R+eJnH/3fB/wQ+IikUyU9D1gEfKnKfn0F+FtJM1Iw/SPw1Yg4eoLtu4DfUlwzGOz+rQMukzRf0jjgHRRz4D8EtlD8tvsmSWMlLQR+97FWSa+U9MwUEL9K2x7rpWvvlDRZ0nSK6wpfTfVPA8skzU77fYqkK6sZa3IzcLmKDwWcAnyQ4gyi26eB5d0XliU1pLFUYyBt7SQ4APLwIYqLnGVvAN5Jcfo+m+JNaCC+THG28QhwIcU0D2lq5OVAC8VvoQ8BH6W4yFetV1PMH+8HbqG4frCxyrafA75A8eme3cD/AW8+0cZpSmM58IM0FTFvsPoXETuB1wD/QnEh/HLg8oh4PCIep7jwuwg4lLa7lSIgAGYC/0lxAXYLcGNEbO6lT+uBu4B7gNuAz6Y+3ELx/K+V9CvgPuDSE+2kwhi2Uzx/aynOBg4DnaV+Xg9soJiqOkxxUfcFVe5+IG3tJHR/isHMRhhJdwKfjojP97NdADPT2d+QSmdVh9Lxdg/18Wxw+QzAbISQ9KeSnpamgFqB5wHfqHW/epJ0uaQnq/hL848D2yg+eWZ1xgFgNnI8G7iX4qO77wD+IiIO1LZLFS2kmO7aTzE11RKeSqhLngIyM8uUzwDMzDI14r/s6cwzz4zm5uZad8PMrK7cddddv4yIht62GfEB0NzcTFtbW627YWZWVyQ92Nc2ngIyM8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8vUiP9LYKsPzUtvq9mx91x3Wc2ObVbPfAZgZpYpB4CZWaYcAGZmmXIAmJllqqoAkHSGpJslPSBph6QXSpoiaaOkXel+cmn7ZZLaJe2UdEmpfqGkbWndDZI0FIMyM7O+VXsGcD3wjYh4DvB8YAewFNgUETOBTekxkmYBLcBsYAFwo6QxaT8rgMUU/4/ozLTezMxqoM8AkDQJeDHwWYCIeDwiDlH8x9Cr02argSvS8kJgbUQciYjdQDswV9I0YFJEbEn/gfSaUhszMxtm1ZwBnAN0AZ+XdLekmySdBkyNiAMA6f6stH0jsK/UviPVGtNyz/pxJC2W1Capraurq18DMjOz6lQTAGOBC4AVEXE+8BhpuucEKs3rRy/144sRKyNiTkTMaWjo9b+0NDOzk1RNAHQAHRFxZ3p8M0UgHEzTOqT7ztL200vtm4D9qd5UoW5mZjXQZwBExEPAPknPTqX5wP3ABqA11VqB9Wl5A9AiabykGRQXe7emaaLDkualT/9cVWpjZmbDrNrvAnoz8CVJpwA/B15HER7rJC0C9gJXAkTEdknrKELiKLAkIo6l/VwLrAImALenm5mZ1UBVARAR9wBzKqyaf4LtlwPLK9TbgHP700EzMxsa/ktgM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsU1UFgKQ9krZJukdSW6pNkbRR0q50P7m0/TJJ7ZJ2SrqkVL8w7add0g2SNPhDMjOzavTnDOClEXFeRMxJj5cCmyJiJrApPUbSLKAFmA0sAG6UNCa1WQEsBmam24KBD8HMzE7GQKaAFgKr0/Jq4IpSfW1EHImI3UA7MFfSNGBSRGyJiADWlNqYmdkwqzYAAviWpLskLU61qRFxACDdn5XqjcC+UtuOVGtMyz3rx5G0WFKbpLaurq4qu2hmZv0xtsrtLoqI/ZLOAjZKeqCXbSvN60cv9eOLESuBlQBz5sypuI2ZmQ1MVWcAEbE/3XcCtwBzgYNpWod035k27wCml5o3AftTvalC3czMaqDPAJB0mqTTu5eBlwP3ARuA1rRZK7A+LW8AWiSNlzSD4mLv1jRNdFjSvPTpn6tKbczMbJhVMwU0FbglfWJzLPDliPiGpB8B6yQtAvYCVwJExHZJ64D7gaPAkog4lvZ1LbAKmADcnm5mZlYDfQZARPwceH6F+sPA/BO0WQ4sr1BvA87tfzfrS/PS22p27D3XXVazY5tZffFfApuZZcoBYGaWKQeAmVmmHABmZplyAJiZZcoBYGaWKQeAmVmmHABmZplyAJiZZcoBYGaWKQeAmVmmHABmZplyAJiZZcoBYGaWKQeAmVmmHABmZplyAJiZZcoBYGaWKQeAmVmmHABmZplyAJiZZcoBYGaWKQeAmVmmHABmZpmqOgAkjZF0t6Rb0+MpkjZK2pXuJ5e2XSapXdJOSZeU6hdK2pbW3SBJgzscMzOrVn/OAN4K7Cg9XgpsioiZwKb0GEmzgBZgNrAAuFHSmNRmBbAYmJluCwbUezMzO2lVBYCkJuAy4KZSeSGwOi2vBq4o1ddGxJGI2A20A3MlTQMmRcSWiAhgTamNmZkNs2rPAD4JvAv4bak2NSIOAKT7s1K9EdhX2q4j1RrTcs/6cSQtltQmqa2rq6vKLpqZWX/0GQCSXgl0RsRdVe6z0rx+9FI/vhixMiLmRMSchoaGKg9rZmb9MbaKbS4CXiXpFcCpwCRJXwQOSpoWEQfS9E5n2r4DmF5q3wTsT/WmCnUzM6uBPs8AImJZRDRFRDPFxd1vR8RrgA1Aa9qsFVifljcALZLGS5pBcbF3a5omOixpXvr0z1WlNmZmNsyqOQM4keuAdZIWAXuBKwEiYrukdcD9wFFgSUQcS22uBVYBE4Db083MzGqgXwEQEZuBzWn5YWD+CbZbDiyvUG8Dzu1vJ83MbPD5L4HNzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMOQDMzDLlADAzy5QDwMwsUw4AM7NMDeR/BLMRqHnpbbXugpnVCZ8BmJllygFgZpYpB4CZWaYcAGZmmXIAmJllygFgZpYpB4CZWab6DABJp0raKuleSdslfTDVp0jaKGlXup9carNMUruknZIuKdUvlLQtrbtBkoZmWGZm1pdqzgCOAC+LiOcD5wELJM0DlgKbImImsCk9RtIsoAWYDSwAbpQ0Ju1rBbAYmJluCwZxLGZm1g99BkAUHk0Px6VbAAuB1am+GrgiLS8E1kbEkYjYDbQDcyVNAyZFxJaICGBNqY2ZmQ2zqq4BSBoj6R6gE9gYEXcCUyPiAEC6Pytt3gjsKzXvSLXGtNyzXul4iyW1SWrr6urqz3jMzKxKVQVARByLiPOAJorf5s/tZfNK8/rRS73S8VZGxJyImNPQ0FBNF83MrJ/69SmgiDgEbKaYuz+YpnVI951psw5geqlZE7A/1Zsq1M3MrAaq+RRQg6Qz0vIE4GLgAWAD0Jo2awXWp+UNQIuk8ZJmUFzs3ZqmiQ5Lmpc+/XNVqY2ZmQ2zar4OehqwOn2S50nAuoi4VdIWYJ2kRcBe4EqAiNguaR1wP3AUWBIRx9K+rgVWAROA29PNzMxqoM8AiIifAOdXqD8MzD9Bm+XA8gr1NqC36wdmZjZM/JfAZmaZcgCYmWXKAWBmlikHgJlZphwAZmaZcgCYmWXKAWBmlikHgJlZphwAZmaZcgCYmWXKAWBmlikHgJlZphwAZmaZcgCYmWXKAWBmlikHgJlZphwAZmaZcgCYmWXKAWBmlikHgJlZphwAZmaZcgCYmWXKAWBmlikHgJlZpvoMAEnTJX1H0g5J2yW9NdWnSNooaVe6n1xqs0xSu6Sdki4p1S+UtC2tu0GShmZYZmbWl2rOAI4C74iI5wLzgCWSZgFLgU0RMRPYlB6T1rUAs4EFwI2SxqR9rQAWAzPTbcEgjsXMzPqhzwCIiAMR8eO0fBjYATQCC4HVabPVwBVpeSGwNiKORMRuoB2YK2kaMCkitkREAGtKbczMbJj16xqApGbgfOBOYGpEHIAiJICz0maNwL5Ss45Ua0zLPetmZlYDVQeApInA14C3RcSvetu0Qi16qVc61mJJbZLaurq6qu2imZn1Q1UBIGkcxZv/lyLi66l8ME3rkO47U70DmF5q3gTsT/WmCvXjRMTKiJgTEXMaGhqqHYuZmfVDNZ8CEvBZYEdEfKK0agPQmpZbgfWleouk8ZJmUFzs3ZqmiQ5Lmpf2eVWpjZmZDbOxVWxzEfBaYJuke1Lt74HrgHWSFgF7gSsBImK7pHXA/RSfIFoSEcdSu2uBVcAE4PZ0MzOzGugzACLiDirP3wPMP0Gb5cDyCvU24Nz+dNDMzIaG/xLYzCxTDgAzs0w5AMzMMlXNReC61bz0tlp3wcxsxPIZgJlZphwAZmaZGtVTQJaHWk317bnuspoc12yw+AzAzCxTDgAzs0w5AMzMMuUAMDPLlAPAzCxTDgAzs0w5AMzMMuUAMDPLlAPAzCxTDgAzs0w5AMzMMuUAMDPLlAPAzCxTDgAzs0w5AMzMMuUAMDPLlAPAzCxTfQaApM9J6pR0X6k2RdJGSbvS/eTSumWS2iXtlHRJqX6hpG1p3Q2SNPjDMTOzalVzBrAKWNCjthTYFBEzgU3pMZJmAS3A7NTmRkljUpsVwGJgZrr13KeZmQ2jPgMgIr4HPNKjvBBYnZZXA1eU6msj4khE7AbagbmSpgGTImJLRASwptTGzMxq4GSvAUyNiAMA6f6sVG8E9pW260i1xrTcs25mZjUy2BeBK83rRy/1yjuRFktqk9TW1dU1aJ0zM7MnnGwAHEzTOqT7zlTvAKaXtmsC9qd6U4V6RRGxMiLmRMSchoaGk+yimZn15mQDYAPQmpZbgfWleouk8ZJmUFzs3ZqmiQ5Lmpc+/XNVqY2ZmdXA2L42kPQV4CXAmZI6gPcD1wHrJC0C9gJXAkTEdknrgPuBo8CSiDiWdnUtxSeKJgC3p5uZmdVInwEQEa8+war5J9h+ObC8Qr0NOLdfvTMzsyHjvwQ2M8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwyNbbWHTCrV81Lb6vZsfdcd1nNjm2jx7CfAUhaIGmnpHZJS4f7+GZmVhjWAJA0BvgUcCkwC3i1pFnD2QczMysM9xTQXKA9In4OIGktsBC4f5j7YVbXajn9VAue8hoawx0AjcC+0uMO4AU9N5K0GFicHj4qaedJHu9M4Jcn2XYk8nhGNo9niOijg7arETOmQdLbeM7uq/FwB4Aq1OK4QsRKYOWADya1RcScge5npPB4RjaPZ+QbbWMa6HiG+yJwBzC99LgJ2D/MfTAzM4Y/AH4EzJQ0Q9IpQAuwYZj7YGZmDPMUUEQclfQm4JvAGOBzEbF9CA854GmkEcbjGdk8npFvtI1pQONRxHFT8GZmlgF/FYSZWaYcAGZmmRqVAVCvXzch6XOSOiXdV6pNkbRR0q50P7m0blka405Jl9Sm15VJmi7pO5J2SNou6a2pXq/jOVXSVkn3pvF8MNXrcjzdJI2RdLekW9Pjeh/PHknbJN0jqS3V6nZMks6QdLOkB9LP0gsHdTwRMapuFBeXfwacA5wC3AvMqnW/quz7i4ELgPtKtX8ClqblpcBH0/KsNLbxwIw05jG1HkOp39OAC9Ly6cBPU5/rdTwCJqblccCdwLx6HU9pXG8HvgzcWs//3krj2QOc2aNWt2MCVgPXpOVTgDMGczyj8Qzgd183ERGPA91fNzHiRcT3gEd6lBdS/CMg3V9Rqq+NiCMRsRtopxj7iBARByLix2n5MLCD4i/B63U8ERGPpofj0i2o0/EASGoCLgNuKpXrdjy9qMsxSZpE8UvhZwEi4vGIOMQgjmc0BkClr5torFFfBsPUiDgAxZsqcFaq1804JTUD51P81ly340nTJfcAncDGiKjr8QCfBN4F/LZUq+fxQBHK35J0V/pKGajfMZ0DdAGfT9N0N0k6jUEcz2gMgKq+bmIUqItxSpoIfA14W0T8qrdNK9RG1Hgi4lhEnEfxF+xzJZ3by+YjejySXgl0RsRd1TapUBsx4ym5KCIuoPjG4SWSXtzLtiN9TGMppoRXRMT5wGMUUz4n0u/xjMYAGG1fN3FQ0jSAdN+Z6iN+nJLGUbz5fykivp7KdTuebuk0fDOwgPodz0XAqyTtoZgmfZmkL1K/4wEgIvan+07gFoopkHodUwfQkc40AW6mCIRBG89oDIDR9nUTG4DWtNwKrC/VWySNlzQDmAlsrUH/KpIkirnLHRHxidKqeh1Pg6Qz0vIE4GLgAep0PBGxLCKaIqKZ4mfk2xHxGup0PACSTpN0evcy8HLgPup0TBHxELBP0rNTaT7FV+cP3nhqfZV7iK6cv4LiUyc/A95d6/70o99fAQ4Av6FI80XAU4FNwK50P6W0/bvTGHcCl9a6/z3G8iKK08+fAPek2yvqeDzPA+5O47kPeF+q1+V4eoztJTzxKaC6HQ/FnPm96ba9+2e/zsd0HtCW/t39OzB5MMfjr4IwM8vUaJwCMjOzKjgAzMwy5QAwM8uUA8DMLFMOADOzTDkAzMwy5QAwM8vU/wNZnvaV05s/gAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(num_orthologs)\n",
    "plt.title(\"Number of orthologs per gene\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "149.9347705300589"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(num_orthologs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "155.0"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.median(num_orthologs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
